kubernetes/websocket/kube_exec.c

```diff
@@ -0,0 +1,122 @@
+#include <stdio.h>
+#include "kube_exec.h"
+
+#define WS_PATH_BUFFER_SIZE 1024
+#define URL_COMMAND_STRING_BUFFER_SIZE 1024
+#define COMMAND_STRING_DELIM " "
+#define COMMAND_PREFIX "&command="
+#define CONTAINER_PREFIX "&container="
+#define ESCAPED_STRING_BUFFER_SIZE 8
+#define ESCAPED_STRING_LENGTH 3
+#define TRUE_STRING "true"
+#define FALSE_STRING "false"
+
+static void default_exec_data_callback(void **p_data_received, long *p_data_received_len)
+{
+    printf("%s: Received %ld bytes:\n%s", __func__, *p_data_received_len, (char *) (*p_data_received));
+}
+
+static void escape_character_in_url(char *escaped, int escaped_buffer_size, const char chr)
+{
+    if (!escaped || escaped_buffer_size < ESCAPED_STRING_LENGTH + 1) {
+        return;
+    }
+
+    if ('+' == chr) {
+        strncpy(escaped, "%2B", ESCAPED_STRING_LENGTH);
+    } else if ('"' == chr) {
+        strncpy(escaped, "%22", ESCAPED_STRING_LENGTH);
+    } else if ('%' == chr) {
+        strncpy(escaped, "%25", ESCAPED_STRING_LENGTH);
+    } else {
+        snprintf(escaped, 2, "%c", chr);
+    }
+}
+
+static int assemble_command_in_url(char *command_string_in_url, int url_command_string_buffer_size, const char *original_command_string)
```

Comment:
If url_command_string_buffer_size is > ESCAPED_STRING_BUFFER_SIZE this will overflow the escaped_string buffer. We should test for this and return -1.

Reply:
The function `escape_character_in_url` is different with `curl_easy_escape` in libcurl.

`ESCAPED_STRING_BUFFER_SIZE` is not the size of the whole escaped string of url, it is the size of string that `only one character` escaped,  e.g. `+` escapes to `%2B` (length is 3) , actually `ESCAPED_STRING_BUFFER_SIZE=4` is enough.

I have test `command_string_length` in function to prevent the command_string_in_url from overflowing 
```c
if (command_string_length >= url_command_string_buffer_size) {
```